// helia/dist/src/utils/libp2p-defaults.js
import { gossipsub } from '@chainsafe/libp2p-gossipsub';
import { noise } from '@chainsafe/libp2p-noise';
import { yamux } from '@chainsafe/libp2p-yamux';
import { bootstrap } from '@libp2p/bootstrap';
import { ipniContentRouting } from '@libp2p/ipni-content-routing';
import { kadDHT } from '@libp2p/kad-dht';
import { mdns } from '@libp2p/mdns';
import { mplex } from '@libp2p/mplex';
import { tcp } from '@libp2p/tcp';
import { webSockets } from '@libp2p/websockets';
import { ipnsSelector } from 'ipns/selector';
import { ipnsValidator } from 'ipns/validator';
import { autoNATService } from 'libp2p/autonat';
import { circuitRelayTransport, circuitRelayServer } from 'libp2p/circuit-relay';
import { identifyService } from 'libp2p/identify';
import { uPnPNATService } from 'libp2p/upnp-nat';
import { bootstrapConfig } from './bootstrappers.js';
export function libp2pDefaults() {
    return {
        addresses: {
            listen: [
                '/ip4/0.0.0.0/tcp/0'
            ]
        },
        transports: [
            tcp(),
            webSockets(),
            circuitRelayTransport({
                discoverRelays: 1
            })
        ],
        connectionEncryption: [
            noise()
        ],
        streamMuxers: [
            yamux(),
            mplex()
        ],
        peerDiscovery: [
            mdns(),
            bootstrap(bootstrapConfig)
        ],
        contentRouters: [
            ipniContentRouting('https://cid.contact')
        ],
        services: {
            identify: identifyService(),
            autoNAT: autoNATService(),
            upnp: uPnPNATService(),
            pubsub: gossipsub(),
            dht: kadDHT({
                validators: {
                    ipns: ipnsValidator
                },
                selectors: {
                    ipns: ipnsSelector
                }
            }),
            relay: circuitRelayServer({
                advertise: true
            })
        }
    };
}
