// libp2p/dist/src/pubsub/dummy-pubsub.js
import { CodeError } from '@libp2p/interfaces/errors';
import { EventEmitter } from '@libp2p/interfaces/events';
import { messages, codes } from '../errors.js';
export class DummyPubSub extends EventEmitter {
    topicValidators = new Map();
    isStarted() {
        return false;
    }
    start() {
    }
    stop() {
    }
    get globalSignaturePolicy() {
        throw new CodeError(messages.PUBSUB_DISABLED, codes.ERR_PUBSUB_DISABLED);
    }
    get multicodecs() {
        throw new CodeError(messages.PUBSUB_DISABLED, codes.ERR_PUBSUB_DISABLED);
    }
    getPeers() {
        throw new CodeError(messages.PUBSUB_DISABLED, codes.ERR_PUBSUB_DISABLED);
    }
    getTopics() {
        throw new CodeError(messages.PUBSUB_DISABLED, codes.ERR_PUBSUB_DISABLED);
    }
    subscribe() {
        throw new CodeError(messages.PUBSUB_DISABLED, codes.ERR_PUBSUB_DISABLED);
    }
    unsubscribe() {
        throw new CodeError(messages.PUBSUB_DISABLED, codes.ERR_PUBSUB_DISABLED);
    }
    getSubscribers() {
        throw new CodeError(messages.PUBSUB_DISABLED, codes.ERR_PUBSUB_DISABLED);
    }
    async publish() {
        throw new CodeError(messages.PUBSUB_DISABLED, codes.ERR_PUBSUB_DISABLED);
    }
}
